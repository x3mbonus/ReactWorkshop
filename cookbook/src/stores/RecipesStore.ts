import { Recipe } from "../models/Recipe";
import { observable, action, computed } from "mobx";
import axios from "axios";

export class RecipesStore {
    @observable recipes: Recipe[] = [];
    
    // @computed get getRecipeById(id:number) : Recipe {
    //     return this.recipes.find(x => x.id == id);
    // }

    @action getRecipeById = (id: number) => {
        // debugger;

        return this.recipes.find(x => x.id == id);
    }

    @action addRecipe = (id:number, title:string, ingredients: string[], description: string) => {
        this.recipes = [...this.recipes, { id: id, title: title, ingredients: ingredients, description: description}];
    }

    @action fetchRecipes = async () => {
            // debugger;
        if (this.recipes.length > 0){
            return;
        }
        const response = await axios.get('/recipes.json');
        this.recipes = response.data;
    }
}