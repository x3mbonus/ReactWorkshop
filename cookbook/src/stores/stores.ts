export enum Stores {
    Recipes = 'RecipesStore'
}