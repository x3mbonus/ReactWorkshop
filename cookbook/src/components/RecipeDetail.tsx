import * as React from 'react';
import { Stores } from "../stores/stores";
import { RecipesStore } from "../stores/RecipesStore";
import { RouteComponentProps } from 'react-router';
import { observer, inject } from 'mobx-react';

interface RecipeDetailRouteParams {
    id: string;
}

export interface IRecipeDetailProps extends RouteComponentProps<RecipeDetailRouteParams> {
    [Stores.Recipes]?: RecipesStore;
}

export interface IRecipeDetailState {
    isLoading: boolean;

}


@inject(Stores.Recipes)
@observer
export default class RecipeDetail extends React.Component<IRecipeDetailProps, IRecipeDetailState> {
  constructor(props: IRecipeDetailProps) {
    super(props);

    
    this.state = {
        isLoading: true
    }
  }

  
    
  async componentDidMount() {
    const {[Stores.Recipes] : recipesStore} = this.props;
    await recipesStore!.fetchRecipes();
    this.setState({
        ...this.state,
        isLoading: false
    })
}



  public render() {

        if (this.state.isLoading) {
            return this.renderText("Loading ...");
        }
        const { [Stores.Recipes]: recipesStore } = this.props;
        const id = parseInt(this.props.match.params.id);
        if (!id){
            return this.renderText("Id is missing");
        }

        const recipe = recipesStore!.getRecipeById(id);
        if (!recipe) {
            return this.renderText("Cannot find recipe");
        }
        let ingredients = recipe.ingredients || [];
        return (
            <div>
                <h3>{recipe.title}</h3>
                <div>{recipe.description}</div>
                <ul>
                    {
                        ingredients!.map((ingredient,i) => 
                            <li key={i}>{ingredient}</li>
                        )
                    }
                </ul>
            </div>
        );
    }

    public renderText = (text:string) => {
        return (
            <div>
                {text}
            </div>
        );
    }
}
