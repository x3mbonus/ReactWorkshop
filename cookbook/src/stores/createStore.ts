import {Stores} from "./stores"
import {RecipesStore} from "./RecipesStore"

export const createStore = () => {
    return {
        [Stores.Recipes]: new RecipesStore
    }
}