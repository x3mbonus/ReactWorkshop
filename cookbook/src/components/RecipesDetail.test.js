import React from "react";
import { RecipesStore } from "../stores/RecipesStore";

import renderer from 'react-test-renderer';

import RecipeDetail from "./RecipeDetail";


it('renders correctly', () => {
    const recipesStore = new RecipesStore;
    recipesStore.recipes = [{
        id: 0,
        title: "Spaghetti with ketchup",
        ingredients: [
            "Spaghetti",
            "Ketchup"
        ],
        description: "Cook spaghetti in boiled water and serve with ketchup."
    },
    {
        id: 1,
        title: "Spaghetti with vodka",
        ingredients: [
            "Spaghetti",
            "Ketchup"
        ],
        description: "Cook spaghetti in boiled water and serve with ketchup."
    }];

    const routeParams = {
        params:{
            id:1
        }
    };

    const tree = renderer.create(<RecipeDetail RecipesStore={recipesStore} match = {routeParams}></RecipeDetail>);

    expect(tree).toMatchSnapshot();
})