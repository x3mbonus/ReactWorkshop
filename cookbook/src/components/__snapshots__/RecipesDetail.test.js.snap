// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <h3>
    Spaghetti with vodka
  </h3>
  <div>
    Cook spaghetti in boiled water and serve with ketchup.
  </div>
  <ul>
    <li>
      Spaghetti
    </li>
    <li>
      Ketchup
    </li>
  </ul>
</div>
`;
